reconciler/leader.go

```diff
@@ -109,9 +115,31 @@ func (laf *LeaderAwareFuncs) Demote(b Bucket) {
 		laf.Lock()
 		defer laf.Unlock()
 		delete(laf.buckets, b.Name())
+		laf.reportBucketCount(len(laf.buckets))
 	}()
 
 	if demote := laf.DemoteFunc; demote != nil {
 		demote(b)
 	}
 }
+
+func (laf *LeaderAwareFuncs) reportBucketCount(count int) {
+	if laf.metricsCtx == nil {
```

Comment:
you have a race here since demote/promote can be independently called in parallel.
I recommend to put it either in ctor, inject it during construction or use a `sync.Once`.

Reply:
We have write locks in promote/demote which can prevent the race?

What's ctor? Even though with construction or `sync.Once`, we need to check whether it is nil every time, right?

Reply:
Not really. After `sync.Once` call, which is the only place where you set the value you can check for it to be nil without mutex, since you know it will always be nil.

Reply:
Added Once.